### PR TITLE
A thread pool

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: base/thread_pool.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+
```
Comment (eggrobin) [inline base/thread_pool.cpp]: Rename to thread_pool_test.cpp

File: base/thread_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    future.wait();
+  }
+
+  bool decreasing = false;
+  for (std::int64_t i = 1; i < numbers.size(); ++i) {
+    if (numbers[i] < numbers[i + 1]) {
+      decreasing = true;
+    }
+  }
+  EXPECT_TRUE(decreasing);
```
Comment (eggrobin) [inline base/thread_pool.cpp]: This test is weird and flaky.

File: base/thread_pool.cpp (hunk 1 above)
Comment (pleroy) [inline base/thread_pool.cpp]: Done.

File: base/thread_pool.cpp (hunk 2 above)
Comment (pleroy) [inline base/thread_pool.cpp]: Fixed so that it tests something, with a minuscule probability of flakiness.

File: base/thread_pool_test.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ protected:
+  ThreadPoolTest() : pool_(std::thread::hardware_concurrency()) {
+    LOG(ERROR) << "Concurrency is " << std::thread::hardware_concurrency();
+  }
+
+  ThreadPool<void> pool_;
+};
+
+// Check that execution occurs in parallel.  If things were sequential, the
+// integers in |numbers| would be non-decreasing.
```
Comment (eggrobin) [inline base/thread_pool_test.cpp]: would be monotone increasing.

File: base/thread_pool_test.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      std::lock_guard<std::mutex> l(lock);
+      numbers.push_back(i);
+    }));
+  }
+
+  for (auto const& future : futures) {
+    future.wait();
+  }
+
+  bool decreasing = false;
```
Comment (eggrobin) [inline base/thread_pool_test.cpp]: monotone_increasing = true

File: base/thread_pool_test.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  for (std::int64_t i = 1; i < numbers.size(); ++i) {
+    if (numbers[i] < numbers[i - 1]) {
+      decreasing = true;
```
Comment (eggrobin) [inline base/thread_pool_test.cpp]: monotone_increasing = false

File: base/thread_pool_test.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+  }
+  EXPECT_TRUE(decreasing);
```
Comment (eggrobin) [inline base/thread_pool_test.cpp]: EXPECT_FALSE(monotone_increasing)

**Pull request merged**